### PR TITLE
add basic `HttpRoundTripper` implementation; minor cleanups

### PR DESCRIPTION
**Pull request by apoelstra**
Takes @sgeisler implementation from #22

### CODE REVIEW

File: src/bitcoind_client.rs
Comment (dpc) [inline src/bitcoind_client.rs:80]: This only works if `reader` is non-blocking? And I can't really tell where in `SimpleBitcoindClient::post` `TcpSocket` is set to non-blocking.

Comment (apoelstra) [reply src/bitcoind_client.rs:80]: Good catch. Yeah, this method looks really weird -- the call to `yield_now` suggests this is supposed to be run in its own thread, but even then, a blocking socket means that the `while` potentially won't ever be reached. There is also a bunch of timeout logic in the code which is used to set `deadline` but which never affects the underlying socket.

Let me see if I can add some unit tests for this.

Comment (apoelstra) [reply src/bitcoind_client.rs:80]: Lol, upon actually testing this there are a ton of ergonomic issues with the API. I will add some more commits.

Comment (apoelstra) [reply src/bitcoind_client.rs:80]: So, for what it's worth, the timeout logic seems to work when testing. I'm just not sure how :)

**Comment by apoelstra** [comment]
Added a bunch of cleanup commits. I have a unit test locally which talks to my bitcoind but I'm not sure how best to commit it, since it currently has my local RPC credentials hardcoded and naturally that won't work for anybody else.

File: src/bitcoind_client.rs
Comment (apoelstra) [reply src/bitcoind_client.rs]: @sgeisler can you offer some insight as to why the timeout code works? I'm really at a loss here, after reading various docs and some of the libstd source I don't see why the first call to `read_line` won't block indefinitely.

Comment (dpc) [reply src/bitcoind_client.rs]: Is some call to https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_nonblocking hiding somewhere?

File: src/simple_rtt.rs
Comment (dpc) [inline src/simple_rtt.rs:131]: Wouldn't `HTTP/1.1` be a better choice? I would imagine that's the most commonly used version, and some servers might just hard-code 1.1 response, instead of downgrading to 1.0 because they don't expect anyone to call with 1.0. Requesting 1.1 and accepting both 1.1 and 1.0 seems like not a lot of complexity with some additional compatibility.

Comment (apoelstra) [reply src/simple_rtt.rs]: Sure, added a commit (and checked that the Liquid integration tests pass when using this to do jsonrpc)

**Review by darosior — APPROVED** [review]
ACK 451f07fefc437215bebe737ccfed016e324e41ad

Tested downstream. Do you plan to do a new release after merging this ?

File: src/simple_rtt.rs
Comment (darosior) [inline src/simple_rtt.rs:138]: Unfortunately, bitcoind will (most of the time...) return both an HTTP and a JSONRPC error on a failed call. However, an error can sometimes be expected and no matter if it was it's always useful to get the JSONRPC as well. So, could we instead read the message till the end even if `!= 200` here and hand it to the caller so that it can decide if it should fail, log, or if it was an expected failure ?

**Comment by apoelstra** [comment]
Yes, the plan is to do a release after merge if I can get a signoff from @stevenroose

**Comment by apoelstra** [comment]
@darosior I added a commit which does not error on a bad HTTP status code. Is this what you were thinking?

I could alternately try to parse the response body, and failing that, return the status code if it was non-200. Rather than always ignoring it.

**Comment by darosior** [comment]
> @darosior I added a commit which does not error on a bad HTTP status code. Is this what you were thinking?
> I could alternately try to parse the response body, and failing that, return the status code if it was non-200. Rather than always ignoring it.

Yeah, i was thinking about the latter but will try out anyways, thanks!

**Comment by apoelstra** [comment]
Ok, added a commit which does the latter.

Also removed a panic which had no right to be there

**Comment by darosior** [comment]
Great, thanks for doing this:
```
[2020-11-09][21:53:21][revaultd::bitcoind::actions][ERROR] Error checking bitcoind: Bitcoind error: Making request: RPC error response: RpcError { code: -28, message: "Loading block index...", data: None }
```
Where i previously had something like "500 internal server error" :D

ACK  328a6f3

File: src/simple_rtt.rs
Comment (dpc) [inline src/simple_rtt.rs:97]: https://docs.rs/http/0.2.1/http/status/struct.StatusCode.html instead of `u16`?

File: src/simple_rtt.rs
Comment (dpc) [inline src/simple_rtt.rs:81]: Isn't `read_line` blocking? I have a deja vu. :D

Comment (apoelstra) [reply src/simple_rtt.rs:81]: lol, that's right, we never did get to the bottom of this

Comment (sgeisler) [reply src/simple_rtt.rs:81]: Sorry I totally forgot about that PR. Some months ago I apparently wrote an answer but did it as a review comment accidentally, so GH never published it :smile: 

> I don't think that works at all in the general case :see_no_evil: even with the underlying stream being nonblocking, the buffered reader would still block until a newline is hit. I never really completed that PR, the main use case was to detect core answering correctly but much to slow (idk why tbh), assuming it always sends complete lines at a time. It's stupid. Nowadays I'd probably just use `async` and timeouts. Or if not possible maybe that would require a custom buffered reader based on a nonblocking underlying stream? The `thread::yield` hack is rather horrible though imo.

> EDIT: I remember, the problem was that core was answering to slowly and I had to wait for lines to arrive (was returning `Ok(0)` a lot) so I built this helper fn, but apparently thought it would be clever to have an option to give up after some time. But it's super unreliable as it assumes Core to send full lines quickly and just take time between them.

Comment (apoelstra) [reply src/simple_rtt.rs:81]: This is even more confusing :). It **does** work, even with a blocking stream and even with `BufRead` adding blocking on top of that blocking. And I didn't notice any slowness, though to be honest I didn't benchmark either.

**Comment by apoelstra** [comment]
Squashed into #36. Further comments should be filed as issues.

**Pull request closed without merging**
